pkg/liquidity-mining/test/L2GaugeCheckpointer.test.ts

```diff
@@ -0,0 +1,223 @@
+import Vault from '@balancer-labs/v2-helpers/src/models/vault/Vault';
+import { actionId } from '@balancer-labs/v2-helpers/src/models/misc/actions';
+import { expect } from 'chai';
+import { ZERO_ADDRESS } from '@balancer-labs/v2-helpers/src/constants';
+import { anyAddressArray } from '@balancer-labs/v2-helpers/src/address';
+
+import { GaugeType } from './GaugeAdder.test';
+
+describe('L2GaugeCheckpointer', () => {
+  let vault: Vault;
+  let adaptor: Contract;
+  let gaugeController: Contract;
+  let gaugeAdder: Contract;
+  let L2GaugeCheckpointer: Contract;
+
+  const gauges = new Map<number, string[]>();
+  let admin: SignerWithAddress;
+
+  let testGaugeType: GaugeType;
+  let testGauges: string[];
+
+  const GAUGES_PER_TYPE = 3;
+  const FIRST_VALID_GAUGE = GaugeType.Polygon;
+
+  // Allowed gauges: Polygon, Arbitrum, Optimism, Gnosis, ZKSync.
+  const GAUGE_TYPES = Object.values(GaugeType)
+    .filter((v) => !isNaN(Number(v)) && v >= FIRST_VALID_GAUGE)
+    .map((t) => Number(t));
+
+  const UNSUPPORTED_GAUGE_TYPES = Object.values(GaugeType)
+    .filter((v) => !isNaN(Number(v)) && v < FIRST_VALID_GAUGE)
+    .map((t) => Number(t));
+
+  before('setup signers', async () => {
+    [, admin] = await ethers.getSigners();
+  });
+
+  before('deploy dependencies: gauge controller and gauge factories', async () => {
+    // Basics: vault, authorizer adaptor and gauge controller.
+    vault = await Vault.create({ admin });
+    adaptor = await deploy('AuthorizerAdaptor', { args: [vault.address] });
+    gaugeController = await deploy('MockGaugeController', { args: [ZERO_ADDRESS, adaptor.address] });
+    // Allow all gauge types in the controller.
+    await Promise.all(GAUGE_TYPES.concat(UNSUPPORTED_GAUGE_TYPES).map(() => gaugeController.add_type('0x', 0)));
+
+    // Gauge factories creation: one per gauge type.
+    const gaugeFactories = await Promise.all(
+      GAUGE_TYPES.map(async (gaugeType) => {
+        return { type: gaugeType, contract: await deploy('MockLiquidityGaugeFactory') };
+      })
+    );
+
+    // Gauge adder & add factories to gauge adder.
+    gaugeAdder = await deploy('GaugeAdder', { args: [gaugeController.address, ZERO_ADDRESS] });
+    const action = await actionId(gaugeAdder, 'addGaugeFactory');
+    await vault.grantPermissionsGlobally([action], admin);
+
+    await Promise.all(
+      gaugeFactories.map((factory) => gaugeAdder.connect(admin).addGaugeFactory(factory.contract.address, factory.type))
+    );
+
+    // Create some gauges from each factory.
+    await Promise.all(
+      gaugeFactories.map(async (factory) =>
+        gauges.set(factory.type, await createGauges(factory.contract, factory.type, GAUGES_PER_TYPE))
+      )
+    );
+  });
+
+  before('get test gauges', () => {
+    testGaugeType = GaugeType.Polygon;
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    testGauges = gauges.get(testGaugeType)!;
+  });
```

Comment:
You can take all of the tests below and put them in a function that receives `gaugeType`, e.g.:

```
fn itAddsAndRemoves(...) {
  describe(`for type ${GaugeType[type]}`, () => {
    ...
  });
}
```

And then call it repeatedly with all types, getting tests for all of them. I'd then have tests outside of this function for invalid types.

Reply:
I started writing the tests with this in mind; at some point I convinced myself it was an overkill.

done :)